Document the NotifyInfo fields for the .NET interop layer.

alonsohki opened the pull request:
The .NET interface for `FileSystemHost.Notify` expects an Action field and Filter field declared as UInt32, making it very cryptic to find out what the values should be. Moreover, the documentation for this can only be found in some issue discussion in the bug tracker.

Added two enumerations that list the possible values for these fields.

billziss-gh commented:
Thank you for the PR. Overall this looks good, but I have some comments:

- AFAIK (I am no C# expert) in order to use an enum constant in C#, one has to use the syntax `EnumName.ConstName`. For example, `NotifyInfoAction.FileActionAdded` and `NotifyInfoFilter.FileNotifyNone`. This looks a bit too verbose and I suggest to change the enums so that the above uses would become `NotifyAction.Added` and `NotifyFilter.None` respectively.

- Some of these action/filter values are not documented, although they can be found in the kernel mode header `ntifs.h`. Should we add the ones that are not documented? (I am open to either way although I note that I do not test the notification API with undocumented values.) The ones that are documented can be found in `winnt.h`:

    ```C
    #define FILE_NOTIFY_CHANGE_FILE_NAME    0x00000001   
    #define FILE_NOTIFY_CHANGE_DIR_NAME     0x00000002   
    #define FILE_NOTIFY_CHANGE_ATTRIBUTES   0x00000004   
    #define FILE_NOTIFY_CHANGE_SIZE         0x00000008   
    #define FILE_NOTIFY_CHANGE_LAST_WRITE   0x00000010   
    #define FILE_NOTIFY_CHANGE_LAST_ACCESS  0x00000020   
    #define FILE_NOTIFY_CHANGE_CREATION     0x00000040   
    #define FILE_NOTIFY_CHANGE_SECURITY     0x00000100   
    #define FILE_ACTION_ADDED                   0x00000001   
    #define FILE_ACTION_REMOVED                 0x00000002   
    #define FILE_ACTION_MODIFIED                0x00000003   
    #define FILE_ACTION_RENAMED_OLD_NAME        0x00000004   
    #define FILE_ACTION_RENAMED_NEW_NAME        0x00000005   
    ```

- A minor nitpick is that your patch appears to contains some stray CRLF line endings. WinFsp source files only use LF line endings (except for files like batch files, etc. where CRLF may be required).

alonsohki commented:
> * AFAIK (I am no C# expert) in order to use an enum constant in C#, one has to use the syntax `EnumName.ConstName`. For example, `NotifyInfoAction.FileActionAdded` and `NotifyInfoFilter.FileNotifyNone`. This looks a bit too verbose and I suggest to change the enums so that the above uses would become `NotifyAction.Added` and `NotifyFilter.None` respectively.
Fixed

> * Some of these action/filter values are not documented, although they can be found in the kernel mode header `ntifs.h`. Should we add the ones that are not documented? (I am open to either way although I note that I do not test the notification API with undocumented values.) The ones that are documented can be found in `winnt.h`:
>   ```c
>   #define FILE_NOTIFY_CHANGE_FILE_NAME    0x00000001   
>   #define FILE_NOTIFY_CHANGE_DIR_NAME     0x00000002   
>   #define FILE_NOTIFY_CHANGE_ATTRIBUTES   0x00000004   
>   #define FILE_NOTIFY_CHANGE_SIZE         0x00000008   
>   #define FILE_NOTIFY_CHANGE_LAST_WRITE   0x00000010   
>   #define FILE_NOTIFY_CHANGE_LAST_ACCESS  0x00000020   
>   #define FILE_NOTIFY_CHANGE_CREATION     0x00000040   
>   #define FILE_NOTIFY_CHANGE_SECURITY     0x00000100   
>   #define FILE_ACTION_ADDED                   0x00000001   
>   #define FILE_ACTION_REMOVED                 0x00000002   
>   #define FILE_ACTION_MODIFIED                0x00000003   
>   #define FILE_ACTION_RENAMED_OLD_NAME        0x00000004   
>   #define FILE_ACTION_RENAMED_NEW_NAME        0x00000005   
>   ```
I will leave that to your discretion. Let me know what you decide and I can update the PR accordingly, if necessary.

> * A minor nitpick is that your patch appears to contains some stray CRLF line endings. WinFsp source files only use LF line endings (except for files like batch files, etc. where CRLF may be required).
fixed

alonsohki commented:
For some reason the tests seem to have failed, but there is no reason for them to fail. Did they get stuck?

billziss-gh commented:
> For some reason the tests seem to have failed, but there is no reason for them to fail. Did they get stuck?

AppVeyor builds are limited to 1.5 hours per job and 3 hours total. Unfortunately testing WinFsp regularly hits these limits now, so I will have to think how to reduce testing and/or make builds faster.

billziss-gh commented:
I have merged this PR in. Thank you @alonsohki for your contribution to this project.

I changed the names of the enums `NotifyInfoFilter` (to `NotifyFilter`) and `NotifyInfoAction` (to `NotifyAction`) as we may introduce an API that uses them without the `NotifyInfo` struct in the future.
